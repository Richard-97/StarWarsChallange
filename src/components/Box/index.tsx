import { View } from 'react-native';
import styled from 'styled-components';
import {
  compose,
  display,
  DisplayProps,
  space,
  SpaceProps,
  position,
  PositionProps,
  flexbox,
  FlexboxProps,
  background,
  BackgroundProps,
  border,
  BorderProps,
  layout,
  LayoutProps,
  BorderRadiusProps,
  boxShadow,
  BoxShadowProps,
  grid,
  GridProps,
} from 'styled-system';

export interface BoxProps
  extends DisplayProps,
    SpaceProps,
    PositionProps,
    FlexboxProps,
    BackgroundProps,
    LayoutProps,
    BorderProps,
    BoxShadowProps,
    GridProps,
    BorderRadiusProps {}

export const Box = styled(View)<BoxProps>`
  ${compose(display, space, position, flexbox, background, border, layout, boxShadow, grid)}
`;
