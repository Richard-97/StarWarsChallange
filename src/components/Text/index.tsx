import styled from 'styled-components/native';
import {
  color,
  ColorProps,
  textAlign,
  TextAlignProps,
  space,
  SpaceProps,
  display,
  DisplayProps,
  layout,
  LayoutProps,
  position,
  PositionProps,
  flex,
  FlexProps,
  fontFamily,
  FontFamilyProps,
} from 'styled-system';
import { COLORS, FONT_FAMILY, FONT_SIZES, FONT_WEIGHTS } from 'src/theme';

interface TextProps
  extends ColorProps,
    TextAlignProps,
    SpaceProps,
    DisplayProps,
    LayoutProps,
    PositionProps,
    FlexProps,
    FontFamilyProps {
  fontSize?: keyof typeof FONT_SIZES;
  fontWeight?: keyof typeof FONT_WEIGHTS;
}

export const Text = styled.Text<TextProps>`
  ${color};
  ${({ fontSize }) => fontSize && `font-size: ${FONT_SIZES[fontSize]};`}
  ${({ fontWeight }) => fontWeight && `font-weight: ${FONT_WEIGHTS[fontWeight]};`}
  ${textAlign};
  ${space};
  ${display};
  ${layout};
  ${position};
  ${flex};
  ${fontFamily};
`;

Text.defaultProps = {
  fontSize: 'M',
  fontWeight: 'NORMAL',
  color: COLORS.BLACK,
  fontFamily: FONT_FAMILY,
};
