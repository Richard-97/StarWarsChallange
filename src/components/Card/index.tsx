import React, { FC, ReactNode } from 'react';
import { Box, BoxProps } from 'src/components/Box';
import { BOX_SHADOWS, COLORS, RADIUS, SPACE } from 'src/theme';

interface Props extends BoxProps {
  children: ReactNode;
}

export const Card: FC<Props> = ({ children, ...boxProps }) => (
  <Box
    p={SPACE.M}
    alignItems="center"
    my={SPACE.XS}
    mx={SPACE.S}
    boxShadow={BOX_SHADOWS.CARD}
    borderRadius={RADIUS.RADIUS_10}
    background={COLORS.TERTIARY}
    {...boxProps}
  >
    {children}
  </Box>
);
