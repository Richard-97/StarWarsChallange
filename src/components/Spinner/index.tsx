import React from 'react';
import { View } from 'react-native';
import RNSpinner, { SpinnerProps, SpinnerType } from 'react-native-spinkit';
import { Box } from 'src/components/Box';
import { COLORS } from 'src/theme';

interface Props extends SpinnerProps {
  isFullScreen?: boolean;
  type?: SpinnerType;
}

export const Spinner: React.FC<Props> = ({ isFullScreen, type, ...spinnerProps }) => {
  if (isFullScreen) {
    return (
      <View>
        <Box height="100%" alignItems="center" justifyContent="center">
          <RNSpinner
            type={type ?? '9CubeGrid'}
            color={COLORS.SECONDARY}
            size={50}
            {...spinnerProps}
          />
        </Box>
      </View>
    );
  }
  return <RNSpinner type={type ?? 'ThreeBounce'} color={COLORS.BLACK} {...spinnerProps} />;
};
