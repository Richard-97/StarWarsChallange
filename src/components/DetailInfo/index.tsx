import React, { FC } from 'react';
import { Text } from 'src/components/Text';
import { Box, BoxProps } from 'src/components/Box';
import { COLORS, RADIUS, SPACE } from 'src/theme';

interface Props extends BoxProps {
  title: string;
  label: string;
}

export const DetailInfo: FC<Props> = ({ title, label, ...boxProps }) => (
  <Box
    p={SPACE.S}
    background={COLORS.SECONDARY}
    borderRadius={RADIUS.RADIUS_10}
    alignItems="center"
    justifyContent="center"
    {...boxProps}
  >
    <Text color={COLORS.TERTIARY} mb={SPACE.XS} fontWeight="SEMI_BOLD">
      {title}
    </Text>
    <Text color={COLORS.TERTIARY}>{label}</Text>
  </Box>
);
