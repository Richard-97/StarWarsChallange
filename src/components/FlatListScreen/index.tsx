import React from 'react';
import styled from 'styled-components/native';
import { FlatList, ListRenderItem, RefreshControl, TouchableOpacity } from 'react-native';
import { Box } from 'src/components/Box';
import { Text } from 'src/components/Text';
import { Spinner } from 'src/components/Spinner';
import { FLAT_LIST_END_REACHED_THRESHOLD } from 'src/constants/common';
import { COLORS, SPACE } from 'src/theme';

const BOTTOM_PADDING = 8;

const Container = styled.SafeAreaView`
  height: 100%;
  background-color: ${COLORS.WHITE};
`;

const DataLoadSpinner = () => (
  <Box alignItems="center">
    <Spinner />
  </Box>
);

interface Props<T> {
  isLoading: boolean;
  canFetchMore: boolean;
  data?: T[];
  keyExtractor?: (item: T) => string;
  itemRenderer: ListRenderItem<T>;
  onEndReached: () => Promise<void>;
  refetch: () => void;
  noDataLabel: string;
}

export const FlatListScreen = <T extends Record<string, any>>({
  noDataLabel,
  isLoading,
  canFetchMore,
  data,
  keyExtractor = ({ id }) => id as string,
  itemRenderer,
  onEndReached,
  refetch,
}: Props<T>) => {
  return (
    <Box>
      <Container>
        {isLoading && !data ? (
          <DataLoadSpinner />
        ) : (
          <FlatList
            overScrollMode="never"
            ListEmptyComponent={
              isLoading ? (
                <DataLoadSpinner />
              ) : (
                <>
                  <Box as={TouchableOpacity} alignItems="center" onPress={() => refetch()}>
                    <Text>Refetching...</Text>
                  </Box>
                  <Text mb={SPACE.M} fontSize="M" textAlign="center">
                    {noDataLabel}
                  </Text>
                </>
              )
            }
            data={data}
            keyExtractor={keyExtractor}
            renderItem={itemRenderer}
            showsVerticalScrollIndicator={false}
            onEndReachedThreshold={FLAT_LIST_END_REACHED_THRESHOLD}
            onEndReached={async () => {
              if (canFetchMore) {
                await onEndReached();
              }
            }}
            {...(isLoading && canFetchMore && { ListFooterComponent: <DataLoadSpinner /> })}
            contentContainerStyle={{ paddingBottom: BOTTOM_PADDING }}
            refreshControl={
              <RefreshControl refreshing={isLoading} onRefresh={refetch} tintColor={COLORS.BLACK} />
            }
          />
        )}
      </Container>
    </Box>
  );
};
