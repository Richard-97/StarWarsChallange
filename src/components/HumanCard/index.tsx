import React, { FC } from 'react';
import { SwapiHuman } from 'src/interfaces/swapi';
import { Text } from 'src/components/Text';
import { Card } from 'src/components/Card';
import { Box } from 'src/components/Box';
import { DetailInfo } from 'src/components/DetailInfo';
import { COLORS, SPACE } from 'src/theme';

export const HumanCard: FC<SwapiHuman> = ({ name, height, mass, birth_year, gender }) => (
  <Card>
    <Text fontSize="XXL" fontWeight="BOLD" mb={SPACE.M} color={COLORS.PRIMARY}>
      {name}
    </Text>
    <Box flexDirection="row">
      <Box flex={1} mr={SPACE.XS}>
        <DetailInfo title="Height" label={height} mb={SPACE.S} />
        <DetailInfo title="Weight" label={mass} />
      </Box>
      <Box flex={1} ml={SPACE.XS}>
        <DetailInfo title="Gender" label={gender} mb={SPACE.S} />
        <DetailInfo title="Birth year" label={birth_year} />
      </Box>
    </Box>
  </Card>
);
