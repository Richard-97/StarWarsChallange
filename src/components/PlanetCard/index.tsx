import React, { FC } from 'react';
import { SwapiPlanet } from 'src/interfaces/swapi';
import { Text } from 'src/components/Text';
import { Card } from 'src/components/Card';
import { Box } from 'src/components/Box';
import { DetailInfo } from 'src/components/DetailInfo';
import { COLORS, SPACE } from 'src/theme';

export const PlanetCard: FC<SwapiPlanet> = ({ name, terrain, climate, gravity, population }) => (
  <Card>
    <Text fontSize="XXL" fontWeight="BOLD" mb={SPACE.M} color={COLORS.PRIMARY}>
      {name}
    </Text>
    <Box flexDirection="row">
      <Box flex={1} mr={SPACE.XS}>
        <DetailInfo title="Terrain" label={terrain} mb={SPACE.S} />
        <DetailInfo title="Climate" label={climate} />
      </Box>
      <Box flex={1} ml={SPACE.XS}>
        <DetailInfo title="Gravity" label={gravity} mb={SPACE.S} />
        <DetailInfo title="Population" label={population} />
      </Box>
    </Box>
  </Card>
);
