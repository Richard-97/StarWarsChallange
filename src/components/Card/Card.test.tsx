import React from 'react';
import { cleanup, render } from '@testing-library/react-native';
import { Card } from 'src/components/Card';
import { Text } from 'src/components/Text';

afterEach(() => {
  cleanup();
});

describe('Card', () => {
  it('matches DOM Snapshot', () => {
    const tree = render(
      <Card data-testId="parent">
        <Text>Test children</Text>
      </Card>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('render children correctly', () => {
    const { getByText } = render(
      <Card data-testId="parent">
        <Text>Test children</Text>
      </Card>
    );

    expect(getByText('Test children')).toBeDefined();
  });
});
