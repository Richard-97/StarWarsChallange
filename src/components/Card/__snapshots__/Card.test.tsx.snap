// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches DOM Snapshot 1`] = `
<View
  alignItems="center"
  background="#201652"
  borderRadius="10px"
  boxShadow="5px 3px 5px #e5e5e5"
  className="sc-bdvvtL iFUkdj"
  data-testId="parent"
  mx="16px"
  my="8px"
  p="24px"
>
  <Text
    color="#000"
    fontFamily="Roboto"
    fontSize="M"
    fontWeight="NORMAL"
    style={
      [
        {
          "color": "#000",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "fontWeight": "400",
        },
      ]
    }
  >
    Test children
  </Text>
</View>
`;
